olmo/optim.py

```diff
@@ -419,21 +429,30 @@ def step(self, closure=None) -> None:
             with torch.enable_grad():
                 closure()
 
-        update_total_dot_prod = torch.tensor(0.0, dtype=torch.float32)
-        update_norms = []
-        signed_update_norms = []
+        update_total_dot_prod: Optional[torch.Tensor] = None
+        update_norms: Optional[List[torch.Tensor]] = None
+        signed_update_norms: Optional[List[torch.Tensor]] = None
+        if self._collecting_metrics and self._record_update_cos_sim:
+            update_total_dot_prod = torch.tensor(0.0, dtype=torch.float32)
+            update_norms = []
+            signed_update_norms = []
 
         for group in self.param_groups:
             for p in group["params"]:
-                if p.grad is None:
+                grad = p.grad
+                if grad is None:
                     continue
 
-                # Perform step weight decay
-                p.data.mul_(1 - group["lr"] * group["weight_decay"])
-
-                grad = p.grad
                 state = self.state[p]
 
+                # Perform step weight decay
+                mask: Optional[torch.Tensor] = None
+                if self._selective_updates:
+                    mask = grad != 0
```

Comment:
thought: you could instead do `mask = grad != 0 if self._selective_updates else 1`, and assume the mask is always present in subsequent logic.

Reply:
good call: 1024122f